dist/images/ovnkube.sh

```diff
@@ -246,24 +246,16 @@ wait_for_event() {
   done
 }
 
-# OVN DBs must be up and initialized before ovn-master and ovn-node PODs can come up
-# This waits for ovnkube-db POD to come up
+# The ovnkbue-db kubernetes service must be populated with OVN DB service endpoints
+# before various OVN K8s containers can come up. This functions checks for that.
 ready_to_start_node() {
-
   # See if ep is available ...
   IFS=" " read -a ovn_db_hosts <<<"$(kubectl --server=${K8S_APISERVER} --token=${k8s_token} --certificate-authority=${K8S_CACERT} \
     get ep -n ${ovn_kubernetes_namespace} ovnkube-db -o=jsonpath='{range .subsets[0].addresses[*]}{.ip}{" "}')"
   if [[ ${#ovn_db_hosts[@]} == 0 ]]; then
     return 1
   fi
   get_ovn_db_vars
-  # cannot use ovsdb-client in the case of raft, since it will succeed even if one of the
```

Comment:
this comment says you need to validate it this way because of raft, but then you added check_ovnkube_db_ep in sb-ovsdb check and not sb-ovsdb-raft. I'm confused. How does ready_to_start_node now ensure the db is up?

Reply:
for the raft, we are checking in from the 

> Do you know why this happens? I'm a little wary of this patch because if we cannot issue the nbctl call it seems the db is not ready, so we shouldn't just skip the check and instead fix the root cause.

This occurs because ovn-nbctl binary tries to download the entire NB DB when trying to execute a subcommand. At scale, this DB is huge. Furthermore, when the ovn-nbctl process starts an inactiivty probe of 5 sec kicks in and if the server doesn't respond within 5sec (possible at scale) for `hello message`, then the connection is reset and the new connection tries to download the database again and we end up in this loop.

There is no need for each of the ovnkube-node daemon in a 1000-node k8s cluster to make a `ovn-nbctl` call to check if it can proceed. It only needs to check if the ovnkube-db service endpoint is populated. Since, whoever populates it already has made that check.

Reply:
according to the docs its a partial DB retrieval, but let's say that is too large, cant you just use the daemon? We obviously make lots ovn-nbctl calls in ovn-k8s and use the daemon to help with perf. I feel like I'm missing something obvious here...

"ovn-nbctl connects to an OVSDB server that
       hosts the northbound database, retrieves a partial copy of the database that  is  complete
       enough  to  do  its  work,  sends  a  transaction  request to the server, and receives and
       processes the server’s reply. In common interactive use, this is fine, but if the database
       is  large, the step in which ovn-nbctl retrieves a partial copy of the database can take a
       long time, which yields poor performance overall.

       To improve performance in such a case, ovn-nbctl offers a "daemon mode,""

Reply:
> this comment says you need to validate it this way because of raft, but then you added check_ovnkube_db_ep in sb-ovsdb check and not sb-ovsdb-raft. I'm confused. How does ready_to_start_node now ensure the db is up?

For the raft, we are checking it in ovndb-raft-functions.sh`ovsdb-raft()

```
  # Create endpoints only if all ovnkube-db pods have started and are running. We do this
  # from the last pod of the statefulset.
  if [[ ${db} == "nb" && "${POD_NAME}" == "ovnkube-db-"${last_node_index} ]]; then
    check_and_apply_ovnkube_db_ep ${port}
  fi
```

Reply:
@trozet all the calls to NB through `ovn-nbctl` happens only on ovnkube-master. We have centralized those calls. These calls go through ovn-nbctld daemon running as a container in the ovnkube-master Pod. This daemon is not available early-on for use in the entrypoint script `ovnkube.sh` for ovn-northd and ovnkube-master containers

We don't make any `ovn-nbctl` calls from the ovnkube-node Pods running on each of the K8s node. So, we don't have ovn-nbctld daemon running in ovnkube-node Pods. Even if we did, it is not going to be available early-on for use in the entrypoint script 'ovnkube.sh' for ovn-controller and db-metrics exporter containers.

Like I said, we don't need those ovn-nbctl calls in there.

Reply:
OK thanks for clarifying. Makes sense.